File: 2024/day11/day11.go
// [Day 11: Plutonian Pebbles](https://adventofcode.com/2024/day/11)

package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

// transform applies a the transformation of a stone according to the process of the puzzle statement.
func transform(stone int) []int {
	// engraved with the number 0 => 1
	if stone == 0 {
		return []int{1}
	}

	// engraved with a number that has an even number of digits => split it into two stones
	stoneStr := strconv.Itoa(stone)
	if len(stoneStr)%2 == 0 {
		mid := len(stoneStr) / 2
		left, _ := strconv.Atoi(stoneStr[:mid])
		right, _ := strconv.Atoi(stoneStr[mid:])
		return []int{left, right}
	}

	// otherwise, multiply by 2024
	return []int{stone * 2024}
}

// blink blinks all stones within the frequency map returning a new frequency map.
func blink(stoneCounts map[int]int) map[int]int {
	newStoneCounts := make(map[int]int)

	for stone, count := range stoneCounts {
		newStones := transform(stone)
		for _, newStone := range newStones {
			newStoneCounts[newStone] += count
		}
	}

	return newStoneCounts
}

// solve reads the input puzzle from data, computes the stone frequency map and blinks it numBlinks times.
// It returns the stone count.
func solve(data []byte, numBlinks int) int {

	strStones := strings.Fields(string(data))
	stoneCounts := make(map[int]int)
	for _, str := range strStones {
		stone, err := strconv.Atoi(str)
		if err != nil {
			return -1
		}
		stoneCounts[stone]++
	}

	for i := 0; i < numBlinks; i++ {
		stoneCounts = blink(stoneCounts)
	}

	totalStones := 0
	for _, count := range stoneCounts {
		totalStones += count
	}

	return totalStones
}

// main is the entry point of the program.
func main() {
	filename := "input.txt"
	if len(os.Args) >= 2 {
		filename = os.Args[1]
	}
	data, err := os.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}

	// part 1
	fmt.Println(solve(data, 25))

	// part 2
	fmt.Println(solve(data, 75))
}

File: 2024/day11/day11.rs
//! [Day 11: Plutonian Pebbles](https://adventofcode.com/2024/day/11)

use rustc_hash::FxHashMap;

struct Puzzle {
    stones: Vec<u64>,
}

impl Puzzle {
    const fn new() -> Self {
        Self { stones: Vec::new() }
    }

    /// Get the puzzle input.
    fn configure(&mut self, data: &str) {
        self.stones = data
            .split_ascii_whitespace()
            .filter_map(|s| s.parse().ok())
            .collect();
    }

    /// Blink all stone within the frequency map according to the blink process:
    /// - stone 0 becomes stone 1
    /// - stone with a even length number splits into two
    /// - otherwise stone number is multiplied by 2024
    fn blink(stone_counts: &FxHashMap<u64, u64>) -> FxHashMap<u64, u64> {
        let mut new_counts = FxHashMap::default();

        for (&stone, &count) in stone_counts {
            let new_stones = {
                // engraved with the number 0 => 1
                if stone == 0 {
                    vec![1]
                } else {
                    let s = stone.to_string();
                    let len = s.len();

                    if len % 2 == 0 {
                        // engraved with a number that has an even number of digits => split it into two stones
                        let mid = len / 2;
                        let left = s[0..mid].parse::<u64>().unwrap();
                        let right = s[mid..].parse::<u64>().unwrap();
                        vec![left, right]
                    } else {
                        // otherwise, multiply by 2024
                        vec![stone * 2024]
                    }
                }
            };

            for new_stone in new_stones {
                *new_counts.entry(new_stone).or_insert(0) += count;
            }
        }

        new_counts
    }

    /// Blink `blinks` times and returns the total number of stones.
    fn solve(&self, blinks: usize) -> u64 {
        let mut stone_counts = FxHashMap::default();

        for &stone in &self.stones {
            *stone_counts.entry(stone).or_insert(0) += 1;
        }

        for _ in 0..blinks {
            stone_counts = Self::blink(&stone_counts);
        }

        stone_counts.values().sum()
    }

    /// Solve part one.
    fn part1(&self) -> u64 {
        self.solve(25)
    }

    /// Solve part two.
    fn part2(&self) -> u64 {
        self.solve(75)
    }
}

fn main() {
    let args = aoc::parse_args();
    let mut puzzle = Puzzle::new();
    puzzle.configure(&args.input);
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}

/// Test from puzzle input
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test01() {
        let mut puzzle = Puzzle::new();
        let data = aoc::load_input_data("test.txt");
        puzzle.configure(&data);
        assert_eq!(puzzle.part1(), 55312);
    }
}
